braindecode/datautil/windowers.py

```diff
@@ -0,0 +1,356 @@
+
+        metadata = {
+            "event_onset_idx": events[:, 0],
+            "trial_number": range(len(events)),
+            "target": events[:, -1],
+        }
+        metadata = pd.DataFrame(metadata)
+        # metadata['subject'] =
+
+        return mne.Epochs(
+            raw,
+            events,
+            tmin=self.trial_start_offset_samples / fs,
+            tmax=(self.window_size_samples - 1) / fs,
+            baseline=None,
+            preload=False,
+            metadata=metadata,
+        )
+
+
+class Windower(object):
+    """
+    A windower that creates a mne Epochs objects. Therefore, it fits supercrops
+    of supercrop_size_samples in trial_start_offset_samples to
+    trial_stop_offset_samples separated by supercrop_stride_samples. If the last
+    supercrop does not end at trial_stop_offset_samples, creates another
+    overlapping supercrop that ends at trial_stop_offset_samples.
+
+    in mne: tmin (s)                    trial onset        onset + duration (s)
+    trial:  |--------------------------------|--------------------------------|
+    here:   trial_start_offset_samples                trial_stop_offset_samples
+
+    Parameters
+    ----------
+    trial_start_offset_samples: int
+        start offset from original trial onsets in samples
+    trial_stop_offset_samples: int
+        stop offset from original trial onsets in samples
+    supercrop_size_samples: int
+        supercrop size
+    supercrop_stride_samples: int
+        stride between supercrops
+    mapping: dict(str: int)
+        mapping from event description to target value
+    """
+    def __init__(self, trial_start_offset_samples, trial_stop_offset_samples,
+                 supercrop_size_samples, supercrop_stride_samples,
+                 drop_samples=False, mapping=None):
+        self.start_offset = trial_start_offset_samples
+        self.stop_offset = trial_stop_offset_samples
+        assert supercrop_size_samples > 0, (
+            "supercrop size has to be larger than 0")
+        self.size = supercrop_size_samples
+        assert supercrop_stride_samples > 0, (
+            "supercrop stride has to be larger than 0")
+        self.stride = supercrop_stride_samples
+        self.drop_samples = drop_samples
+        self.mapping = mapping
+
+    # TODO: handle case we don't get a raw
+    def __call__(self, raw):
+        # EventWindower: if annotations are given in raw, use them to get onsets
+        assert hasattr(raw, "annotations") and hasattr(raw.annotations, "onset")
+        fs = raw.info["sfreq"]
+        onsets = round_list_to_int(raw.annotations.onset*fs)
+        i_trials, starts, stops = _supercrop_starts(
+            onsets, self.start_offset, self.stop_offset, self.size, self.stride,
+            self.drop_samples)
+        description = raw.annotations.description
+        if self.mapping is not None:
+            description = [self.mapping[d] for d in description]
```

Comment:
Here first, description is an array of str. If mapping is set, description will become a list of int.

If mapping is set by the user, e.g. to only have specific subset of events, the mapping will not contain every description from the annotations object. This will probably lead to an error.

On the other hand, this is good to transform the str descriptions to event codes, but then mapping should be mandatory I think.